### PR TITLE
Python SK: Simplify memory_store_base

### PR DESCRIPTION
**Pull request by awharrison-28**
### Motivation and Context
This PR significantly reduces the complexity of implementing a new memory store. Previously, memory stores needed to implement EmbeddingIndexBase, DataStoreBase, and MemoryStoreBase. Now, memory stores only need to implement MemoryStoreBase. This PR also brings MemoryStoreBase to parity with the .NET IMemoryStore interface.  

### Description
- Removed `data_entry.py`, `data_store_base.py`, `volatile_data_store.py`, and `embedding_index_base.py`
- Added `with_embedding` parameter to all get/search methods. Passing the embedding between the application and storage is not optional and defaults to False
- `MemoryStoreBase` (Python SK) == `IMemoryStore` (.NET SK)
- Updated `volatile_memory_store` to implement `MemoryStoreBase`
- Removed `ABC` from `MemoryStoreBase` 
- Added documentation to memory files

Closes https://github.com/microsoft/semantic-kernel/issues/670

### Contribution Checklist

- [x] The code builds clean without any errors or warnings
- [x] The PR follows SK Contribution Guidelines (https://github.com/microsoft/semantic-kernel/blob/main/CONTRIBUTING.md)
- [x] The code follows the .NET coding conventions (https://learn.microsoft.com/dotnet/csharp/fundamentals/coding-style/coding-conventions) verified with `dotnet format`
- [x] All unit tests pass, and I have added new tests where possible
- [x] I didn't break anyone :smile:

### CODE REVIEW

File: python/semantic_kernel/memory/memory_store_base.py [hunk 1]
```diff
@@ -1,12 +1,120 @@
...
+        embedding: ndarray,
+        limit: int,
+        min_relevance_score: float,
+        with_embeddings: bool
+        # TODO: cancel token
+    ) -> List[Tuple[MemoryRecord, float]]:
+        pass
+
+    @abstractmethod
+    async def get_nearest_matche_async(
```
Comment (lemillermicrosoft) [inline python/semantic_kernel/memory/memory_store_base.py:112]: nit:
```suggestion
    async def get_nearest_match_async(
```

**Review by lemillermicrosoft — APPROVED** [review]

File: python/semantic_kernel/memory/volatile_memory_store.py [hunk 2]
```diff
@@ -1,37 +1,300 @@
...
+        """Initializes a new instance of the VolatileMemoryStore class.
+
+        Arguments:
+            logger {Optional[Logger]} -- The logger to use. (default: {None})
+        """
+
+    async def create_collection_async(
+        self,
+        collection_name: str
+        # TODO: cancel token
```
Comment (shawncal) [inline python/semantic_kernel/memory/volatile_memory_store.py:30]: Trying to keep todos out of the code as much as possible. If you add these, can you please add an Issue to the github board, and add the task number to the TODO comment?

File: python/semantic_kernel/memory/memory_store_base.py [hunk 3]
```diff
@@ -1,12 +1,120 @@
...
 
-class MemoryStoreBase(DataStoreBase, EmbeddingIndexBase, ABC):
-    pass
+
+class MemoryStoreBase:
+    @abstractmethod
+    async def create_collection_async(
+        self,
+        collection_name: str
+        # TODO: cancel token
```
Comment (alexchaomander) [inline python/semantic_kernel/memory/memory_store_base.py:16]: Do we need all these TODOs? If so, it's unclear to me what someone would need to do to address it

**Review by alexchaomander — APPROVED** [review]
Looks good to me! I think just need to remove the unnecessary TODOs

File: python/semantic_kernel/memory/volatile_memory_store.py (hunk 2 above)
Comment (awharrison-28) [reply python/semantic_kernel/memory/volatile_memory_store.py]: removed TODOs for cancellation tokens and created a task to track the work

File: python/semantic_kernel/memory/memory_store_base.py (hunk 3 above)
Comment (awharrison-28) [reply python/semantic_kernel/memory/memory_store_base.py]: removed the todos and created a task to track the work

**Review by alexchaomander — DISMISSED** [review]

**Comment by lemillermicrosoft** [comment]
Looks like there are some lint failures to address.

**Review by lemillermicrosoft — APPROVED** [review]

**Pull request merged**